src/hotspot/share/services/heapDumper.cpp

```diff
@@ -2291,17 +2283,19 @@ void VM_HeapDumper::work(uint worker_id) {
       LockedClassesDo locked_load_classes(&do_load_class);
       ClassLoaderDataGraph::classes_do(&locked_load_classes);
     }
-    Universe::basic_type_classes_do(&do_load_class);
 
     // write HPROF_FRAME and HPROF_TRACE records
     // this must be called after _klass_map is built when iterating the classes above.
     dump_stack_traces();
 
     // Writes HPROF_GC_CLASS_DUMP records
+    // For array classes we need signers and protection domain from their bottom classes
+    // so do_class_dump skips array classes and dumps array classes with corresponding instance class.
```

Comment:
I'm having trouble parsing this comment. Can you clarify?

Reply:
HPROF_GC_CLASS_DUMP records contain references to signers and protection domain objects.
But this is properties of InstanceKlass, for array classes the properties from their element classes are written (or NULL for arrays of primitive types).
do_class_dump callback skips array classes:
  if (k->is_instance_klass()) {
    DumperSupport::dump_class_and_array_classes(writer(), k);
  }

dump_class_and_array_classes dumps the class and then iterate over array classes (k->array_klass_or_null()) and writes records for them (using signer and protection domain from the instance class specified).

This approach also skips primitive type arrays (they don't have corresponding InstanceKlass) - they are dumped separately.

I added the comment to explain why do_class_dump goes this tricky way.
Do you have any suggestions how to rephrase it to make clearer?

Reply:
Sorry about the delay. I didn't see your response when it came in.

It seems like this comment belongs with do_class_dump(). Putting it here is kind of confusing. Also, what is meant by "bottom class"? Do you mean the element type of the array?

Reply:
This comment explains why we call ClassLoaderDataGraph::classes_do(do_class_dump) and then Universe::basic_type_classes_do(&do_basic_type_array_class_dump).

Bottom klass is 1-dimensional type.
For example for [[LObject; element type is [lObject;, and bottom klass is LObject;
Looked at the code and see that ObjAyyaKlass actually contains link to it - see ObjArrayKlass::bottom_klass(), so I can update do_class_dump to use it.

Reply:
I updated the fix, fixed do_class_dump and removed this comment.
This also removes code duplication in dump_class_and_array_classes and dump_basic_type_array_class